### PR TITLE
fix #155 refactor(database): standardize Hall symbols

### LINKED ISSUE #155: Unconventional Hall symbols in `spg_database.c`
I am not sure if this is a bug or not. But, I have noticed some Hall symbols in `spg_database.c` differ from the list of Hall symbols in International Tables for Crystallography Vol. B (ITB).

In `spglib`, the so-called "V entry" in Hall symbols is not consistent with ITB: For example, in Table A1.4.2.7 of ITB, No. 151 is represented as `P 31 2 (0 0 4)`. On the other hand, `spg_database.c` represents it as `P 31 2c (0 0 1)`.
https://github.com/spglib/spglib/blob/188e6a5aec5424ce5bbdeebc98502439a4ae3713/src/spg_database.c#L480

### PR DESCRIPTION
**Pull request by site-g**
I updated the 3D space group Hall symbols to be the same as the ones in ITB (2010). I also modified the layer group Hall symbols. Now every layer group Hall symbol (except 20, 52:1, 54 and 64) has a space group counterpart in ITB. They share the same symbol, except we use lowercase 'p' and 'c' to represent layer group. The four symbols are different because the origin choice are different in ITA and ITE, as shown in the following table and diagram. The layer group Hall symbol table is added to Appendix D in the overleaf.
| Space group number | Space H-M | Space Hall | Layer group number | Layer H-M | Layer Hall |
| - | - | - | - | - | - |
| 17:cab | P 21 2 2 | P 2a 2a | 20 | p 21 2 2 | p 2 2a |
| 85:1 | P 4/n:1 | P 4ab -1ab | 52:1 | p 4/n:1 | p 4 -1ab |
| 90 | P 4 21 2 | P 4ab 2ab | 54 | p 4 21 2 | p 4 2ab |
| 129:1 | P 4/n m m:1 | P 4ab 2ab -1ab | 64:1 | p 4/n m m:1 | p 4 2ab -1ab |
p4n_compare

@atztogo, @lan496 could you take some time to double check this as there are too many entries and is very error-prone.

Fixes #155

### CODE REVIEW

**Comment by LecrisUT** [comment]
Hmm, I think we should add the database generation to the CI, can someone point me to the steps for that? The database folder could also use some refactoring

**Comment by site-g** [comment]
I'm not very familar with CI. Do you mean you want to generate the database automatically? I think it is a little difficult.
The step is:

1. Manually write the *.csv files. They are generated from International Tables for Crystallography. (I used excel to generate it.)
2. Use the *.csv files as inputs. Input them into appropriate *.py programs to get the output.
3. Copy the output to the correpsonding lines of src/*_database.c.

For example,
```
python hall2operations.py --dump layer_spg.csv
    0,        /* dummy */
    16484,    /*    1 (  1) [ 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0] */
    16484,    /*    2 (  2) [ 1, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0] */
    3198,     /*    3 (  2) [-1, 0, 0, 0,-1, 0, 0, 0,-1, 0, 0, 0] */
......
    3119,     /*  756 (116) [-1, 0, 0,-1, 1, 0, 0, 0, 1, 0, 0, 0] */
    14133,    /*  757 (116) [ 1,-1, 0, 0,-1, 0, 0, 0,-1, 0, 0, 0] */
    5549,     /*  758 (116) [-1, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0] */
```
The output is just https://github.com/spglib/spglib/blob/9957bfc15e94bc5d1a495d43fb391f0201897f52/src/spg_database.c#L8495-L9255
after some modification.

In fact, the database does not need to be generated by the user. Do we need to add it to the CI?

**Comment by LecrisUT** [comment]
Yes it should be added to check PRs like these. What kind of manual editing do you do?

> In fact, the database does not need to be generated by the user. Do we need to add it to the CI?

Exactly because of that it should be in the CI. Like a pre-commit.

**Comment by site-g** [comment]
Ok, then this PR only used `hall2operations.py`,  `make_spgtype_db.py`, `spg.csv` and `layer_spg.csv`. Modification appears because these programs are not written for layer groups.
For space group
`python hall2operations.py --dump spg.csv` gives all hall symbols' symmetry operations in `static const int symmetry_operations[]`
 https://github.com/spglib/spglib/blob/9957bfc15e94bc5d1a495d43fb391f0201897f52/src/spg_database.c#L1107-L8495
`python hall2operations.py --ref spg.csv` gives the range of each hall symbols' symmetry operations `static const int symmetry_operation_index[][2]`, i.e. {num_operations, start_index} of each group in `symmetry_operations[]` https://github.com/spglib/spglib/blob/9957bfc15e94bc5d1a495d43fb391f0201897f52/src/spg_database.c#L9258-L9788
These two output can be directly copied to spg_database.c after code formating.
Then for layer group, the symmetry operations are added after the ones of space groups in `symmetry_operations[]`. So the start index need to be shifted. https://github.com/spglib/spglib/blob/9957bfc15e94bc5d1a495d43fb391f0201897f52/database/hall2operations.py#L344
and https://github.com/spglib/spglib/blob/9957bfc15e94bc5d1a495d43fb391f0201897f52/database/hall2operations.py#L420 need to be change to `count = 7388`. 
`python hall2operations.py --dump layer_spg.csv` gives the rest of `symmetry_operations[]` https://github.com/spglib/spglib/blob/9957bfc15e94bc5d1a495d43fb391f0201897f52/src/spg_database.c#L8497-L9254
`python hall2operations.py --ref layer_spg.csv` gives `static const int layer_symmetry_operation_index[][2]` https://github.com/spglib/spglib/blob/9957bfc15e94bc5d1a495d43fb391f0201897f52/src/spg_database.c#L10030-L10146

`make_spgtype_db.py` outputs `static const SpacegroupType spacegroup_types[]` 
https://github.com/spglib/spglib/blob/9957bfc15e94bc5d1a495d43fb391f0201897f52/src/spg_database.c#L40-L1103
However, this code does not support layer group now. So I generate https://github.com/spglib/spglib/blob/9957bfc15e94bc5d1a495d43fb391f0201897f52/src/spg_database.c#L9791-L10027
manually (excel again).

If you want to automate it. An `if` clause should be added for `count` in `hall2operations.py`. `make_spgtype_db.py` need to be rewritten to support layer group. I did not use other python scripts in this PR. These codes were written 10 years ago, I'm not sure if other scripts could work properly now.

**Comment by LecrisUT** [comment]
> If you want to automate it. An `if` clause should be added for `count` in `hall2operations.py`. `make_spgtype_db.py` need to be rewritten to support layer group. I did not use other python scripts in this PR. These codes were written 10 years ago, I'm not sure if other scripts could work properly now.

For the most part they should still work, but they really need to be updated:

Screenshot_20230821_152521

Best is to just migrate them into the already provided python package and fix the issues there. Other than that there are no manual formatting issues or anything else?

**Comment by site-g** [comment]
> Other than that there are no manual formatting issues or anything else?

I'm not sure. One thing is that some of the codes are written in python 2. Did that cause the errors?

File: database/hall2operations.py [hunk 1]
```diff
@@ -333,10 +333,11 @@ def _translation(self, N):
...
     def _change_of_basis_symbol(self, V):
-        if V[0] == '-':
-            return np.array([0, 0, -1])
-        else:
-            return np.array([0, 0, 1])
+        # if V[0] == '-':
+        #     return np.array([0, 0, -1])
+        # else:
+        #     return np.array([0, 0, 1])
+        return np.array([0, 0, int(V[:-1])])
```
Comment (LecrisUT) [inline database/hall2operations.py:336]: Please don't comment code, just replace it

Comment (site-g) [reply database/hall2operations.py:336]: I have removed them.

File: database/hall2operations.py [hunk 2]
```diff
@@ -333,10 +333,7 @@ def _translation(self, N):
         return T
 
     def _change_of_basis_symbol(self, V):
-        if V[0] == '-':
-            return np.array([0, 0, -1])
-        else:
-            return np.array([0, 0, 1])
```
Comment (lan496) [inline database/hall2operations.py:339]: @atztogo Why does the original implementation ignore the value of `V`?
For example, this function returns `[0, 0, 1]` for hall symbol `"P 62 2 (0 0 4) "`.

**Comment by lan496** [comment]
I have confirmed spg.csv and layer_spg.csv.
I would like to note that this PR includes more than https://github.com/spglib/spglib/issues/155: for Hall numbers 494-502 (ITA No. 200-206), three-fold rotations are corrected to roto inversion (e.g. Pm3m -> Pm-3m)

File: database/hall2operations.py (hunk 2 above)
Comment (site-g) [reply database/hall2operations.py:339]: Because the original implementation is compatible with the original spg.csv, which only has (0 0 1) and (0 0 -1).

Comment (LecrisUT) [reply database/hall2operations.py:339]: There is actually no need to keep compatibility with old `spg.csv` and such. These are all version controlled together

Comment (site-g) [reply database/hall2operations.py:339]: @LecrisUT I think lan496 was asking why the old version `hall2operations.py` cannot handle something like `np.array([0, 0, 4])`, and I was explaining that the old version does not need to handle this. The PR does not deliberately keep the compatibility. Now the new `spg.csv` contains both, so the code need the ability to handle both.

Comment (atztogo) [reply database/hall2operations.py:339]: @lan496, my memory has completely gone about Hall symbols. But I think the answer by @site-g is explaining it.

Comment (lan496) [reply database/hall2operations.py:339]: I see. Thank you for your explanations!

**Review by lan496 — APPROVED** [review]

**Comment by atztogo** [comment]
If this PR changes the output of the Hall symbol, what is the best way to inform it to users? I think writing it in the release not is enough though, but not just as a commit log.

**Comment by lan496** [comment]
I think a description in the release note will be the best way.

**Review by atztogo — APPROVED** [review]

**Comment by atztogo** [comment]
Thanks @site-g, @lan496, @LecrisUT.

**Pull request merged**